### PR TITLE
Ignore-quicker-pending-drop-token

### PR DESCRIPTION
**Pull request by haixuanTao**
This PR reduces the time required to ignore unsent drop token from 30s to 1s as in Python this error is linked to a drop token race condition with the GIL.

### CODE REVIEW

**Comment by haixuanTao** [comment]
CI is blocked by: https://community.anaconda.cloud/t/cant-download-anaconda-say-violating-terms-of-service/76074/14

File: binaries/daemon/src/lib.rs [hunk 1]
```diff
@@ -1567,7 +1567,7 @@ impl RunningDataflow {
         let running_nodes = self.running_nodes.clone();
         let grace_duration_kills = self.grace_duration_kills.clone();
         tokio::spawn(async move {
-            let duration = grace_duration.unwrap_or(Duration::from_millis(500));
+            let duration = grace_duration.unwrap_or(Duration::from_millis(2000));
```
Comment (phil-opp) [inline binaries/daemon/src/lib.rs:1570]: This change is not really related to this PR, is it? It's probably better to do it in a separate PR so users can see that the default was changed in the changelog.

**Review by phil-opp — COMMENTED** [review]
> This PR reduces the time required to ignore unsent drop token from 30s to 1s as in Python this error is linked to a drop token race condition with the GIL.

Could you clarify how this happens? It sounds like the cleaner solution would be to avoid this race condition and ensure that drop messages are properly sent.

**Comment by haixuanTao** [comment]
So, @phil-opp if you want to reproduce the race condition, you can use the following example of a dataflow based on the `python-dataflow` example:

```yaml
nodes:
  - id: webcam
    custom:
      source: ./webcam.py
      inputs:
        tick:
          source: dora/timer/millis/50
          queue_size: 1000
      outputs:
        - image

  - id: object_detection
    custom:
      source: ./object_detection.py
      inputs:
        image: webcam/image
      outputs:
        - bbox

  - id: object_detection_1
    custom:
      source: ./object_detection.py
      inputs:
        image: webcam/image
      outputs:
        - bbox
  - id: object_detection_2
    custom:
      source: ./object_detection.py
      inputs:
        image: webcam/image
      outputs:
        - bbox
  - id: object_detection_3
    custom:
      source: ./object_detection.py
      inputs:
        image: webcam/image
      outputs:
        - bbox
  - id: object_detection_4
    custom:
      source: ./object_detection.py
      inputs:
        image: webcam/image
      outputs:
        - bbox
  - id: plot
    custom:
      source: ./plot.py
      inputs:
        image: webcam/image
        bbox: object_detection/bbox
  - id: plot_1
    custom:
      source: ./plot.py
      inputs:
        image: webcam/image
        bbox: object_detection/bbox

  - id: plot_2
    custom:
      source: ./plot.py
      inputs:
        image: webcam/image
        bbox: object_detection/bbox

  - id: plot_3
    custom:
      source: ./plot.py
      inputs:
        image: webcam/image
        bbox: object_detection/bbox

```

By starting, waiting couple of seconds and sending a stop signal either by using ctl-c or stop command the nodes are not able to exit within the grace duration, and exit after 30s for the timeout of drop token to kick in.

```bash

[ERROR]
Dataflow 0190772a-f3e5-75a6-b7db-ddce5fad5a4e failed:

Node `object_detection` failed: node was killed by dora because it didn't react to a stop message in time (SIGKILL)
Node `object_detection_1` failed: node was killed by dora because it didn't react to a stop message in time (SIGKILL)
Node `object_detection_3` failed: node was killed by dora because it didn't react to a stop message in time (SIGKILL)
Node `object_detection_4` failed: node was killed by dora because it didn't react to a stop message in time (SIGKILL)
Node `object_detection_2` failed: node was killed by dora because it didn't react to a stop message in time (SIGKILL)
```

I have tried many, many things to solve it and so far, the only way I can make a graceful exit is by reducing the timeout on those receivers.

I would love to have a better solution, but haven't figured out any.

File: apis/rust/node/src/node/mod.rs [hunk 2]
```diff
@@ -401,7 +401,7 @@ impl Drop for DoraNode {
                 );
             }
 
-            match self.drop_stream.recv_timeout(Duration::from_secs(10)) {
+            match self.drop_stream.recv_timeout(Duration::from_millis(500)) {
```
[the remark this answers is not in the document]
Comment (haixuanTao) [reply apis/rust/node/src/node/mod.rs:404]: It might be but this only happen on a Node Drop. In the context of stopping a dataflow, so even though this might create read error, it is only in the context of dropping nodes.

Also, I think that dropping node should be fast and having to wait 10 seconds is never something that should be acceptable.

File: binaries/daemon/src/lib.rs (hunk 1 above)
Comment (haixuanTao) [reply binaries/daemon/src/lib.rs:1570]: Sure.

File: apis/rust/node/src/node/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (phil-opp) [reply apis/rust/node/src/node/mod.rs:404]: It would also happen when a node finishes normally and the other nodes keep running, wouldn't it?

I agree that a 10s _in the drop handler_ is too long. But we should still wait the 10s before exiting the process and removing the shared memory that might still be needed by receiving nodes. Alternatively, we could to transfer ownership of the shared memory region to e.g. the deamon to enable a node exit without wait.
